### PR TITLE
Create a password strength UI for security dashboards plugin

### PR DESCRIPTION
**Pull request by derek-ho**
### Description
Adds a UX for the password strength feature added to the security plugin backend. It adds a bar that shows progress that updates itself as the password is entered, giving real time feedback to the user. Without this it can be frustrating to try different passwords without feedback on how it is weak, or would the password be accepted by the backend as strong enough.

### Category
New feature

### Why these changes are required?
UX to support password strength feature from the backend

### What is the old behavior before changes and new behavior after changes?
Old: no password strength bar for both create/edit/duplicate user page and the self reset panel
New: Password strength bar for both the page and the panel

https://github.com/opensearch-project/security-dashboards-plugin/assets/26328171/93fbe33e-7e12-4fec-8ff0-6fc1d9cf8d73

### Issues Resolved
Fix: #1523 

### Testing
Manual testing, unit tests

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by RyanL1997** [comment]
Can we have like a preview screenshot for this?

File: public/apps/configuration/utils/password-strength-bar.tsx
Comment (DarshitChanpura) [inline public/apps/configuration/utils/password-strength-bar.tsx:27]: is there a message provided by zxcvbn that we can use directly?

**Review by DarshitChanpura — COMMENTED** [review]
thanks for adding this @derek-ho . Could you add some screen-shots to visualize the final look? Also could you please update the branch to pull latest changes from main,

**Comment by derek-ho** [comment]
> Can we have like a preview screenshot for this?

Added a video snippet to the PR description: https://github.com/opensearch-project/security-dashboards-plugin/pull/1762#issue-2108188423

File: public/apps/configuration/utils/password-strength-bar.tsx
Comment (derek-ho) [reply public/apps/configuration/utils/password-strength-bar.tsx:27]: The score is simply a value 0-4 indicating the strength of the password. Zxcvbn also provides a suggestions and warnings if applicable, which I have also added - see below in the code and you can also check it out in the video.

**Comment by derek-ho** [comment]
> thanks for adding this @derek-ho . Could you add some screen-shots to visualize the final look? Also could you please update the branch to pull latest changes from main,

Done in: [72a63c0](https://github.com/opensearch-project/security-dashboards-plugin/pull/1762/commits/72a63c0ebd8f0446b06ffb61a9f6670ea47cd906), and updated the PR description with a video

**Comment by scrawfor99** [comment]
@joshuarrrr could you please confirm that this looks good from the OUI team. @Naarcha-AWS or 
@kolchfa-aws can you give input on the wording for this.

File: public/apps/account/password-reset-panel.tsx
Comment (scrawfor99) [inline public/apps/account/password-reset-panel.tsx:22]: Should we be using the OUI component library?

File: public/apps/account/password-reset-panel.tsx
Comment (scrawfor99) [inline public/apps/account/password-reset-panel.tsx:39]: Should the exported component follow lowercase naming conventions?

**Review by scrawfor99 — APPROVED** [review]
Looks good to me, thanks Derek

**Comment by kolchfa-aws** [comment]
@scrawfor99 LGTM

File: public/apps/account/password-reset-panel.tsx
Comment (derek-ho) [reply public/apps/account/password-reset-panel.tsx:22]: There is actually a deep link from Eui to Oui - see here: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/package.json#L139, I guess this is due to legacy reasons and refactoring may be a pain.

File: public/apps/account/password-reset-panel.tsx
Comment (derek-ho) [reply public/apps/account/password-reset-panel.tsx:39]: Since this is a component and not a function the naming convention is to use an upper case. The above examples are for helper functions, which would use camel case.

File: public/apps/configuration/utils/test/password-strength-bar.test.tsx
Comment (DarshitChanpura) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: IMO: `This is a top-100 common password` seems like an extra information. I would like to keep it minimal and relevant.

Comment (DarshitChanpura) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: okay, I have no more questions. Looks good to me.

File: public/apps/configuration/utils/password-strength-bar.tsx
Comment (RyanL1997) [inline public/apps/configuration/utils/password-strength-bar.tsx:31]: nit: may need to add i18n support for all these msgs. But we can do that as a follow-up.

File: public/apps/configuration/utils/test/password-strength-bar.test.tsx
Comment (RyanL1997) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: I was more wondering, if there is a case that the password is not an valid password (e.g. too short / missing a special character ... ), then what should we show here? Or all these msg should only be popped up when the password format meets the minimum requirements?

**Review by RyanL1997 — COMMENTED** [review]
Hi @derek-ho , thanks for doing this enhancement. In generally this is LGTM.

File: public/apps/configuration/utils/test/password-strength-bar.test.tsx
Comment (cwperks) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: This is an example I like from Wix: https://miro.medium.com/v2/resize:fit:1400/format:webp/1*v3N8tIHyKPa3Pz7lSnpPiA.gif

See the full article here: https://wix-ux.com/how-we-increased-our-users-passwords-strength-by-26-ab25efffa622

I've never seen a password set/reset UX where it tells me the password is too common. I've only seen UX that tells me the pw is too weak. Sometimes a site may also provide the minimum checklist if the website has explicit restrictions on character length, numbers and special characters. 

The warnings I think would be appropriate is if the user has CAPS LOCK on inadvertently.

Comment (derek-ho) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: I see. Yea from what I've seen in their source code, I think suggestions seem more applicable to helping the user generate a stronger password. The only useful warning I've seen is "Common names and surnames are easy to guess", but otherwise suggestions should help more - @DarshitChanpura @cwperks I'll just remove the warnings block and keep the suggestions?

File: public/apps/configuration/utils/password-strength-bar.tsx
Comment (derek-ho) [reply public/apps/configuration/utils/password-strength-bar.tsx:31]: Sure, let me create an issue. I think there is a decent amount of text across the repo that is not translated. However, I do not know how to replicate this/change the OSD to use a different default language, so I may need to look into that.

Comment (derek-ho) [reply public/apps/configuration/utils/password-strength-bar.tsx:31]: Created issue here: https://github.com/opensearch-project/security-dashboards-plugin/issues/1886

File: public/apps/configuration/utils/test/password-strength-bar.test.tsx
Comment (derek-ho) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: @RyanL1997 regarding your question [here](https://github.com/opensearch-project/security-dashboards-plugin/pull/1762#discussion_r1560073276), that is actually a separate issue - see: https://github.com/opensearch-project/security-dashboards-plugin/issues/1234. Since the backend has two requirements, one which is regex matching, and one which is password strength via `zxcvbn` library, we need to add UX to satisfy both. This is just for `zxcvbn` library, and I will hopefully create another PR to add UX surrounding the regex requirements. However, I think it would be fair to say that _most_ of the time, if they satisfy strong password via `zxcvbn` they will also satisfy regex.

Comment (derek-ho) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: @kamingleung can you help chime in here for final path forward? Thanks!

Comment (kamingleung) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: As I mentioned in https://github.com/opensearch-project/security-dashboards-plugin/issues/1523#issuecomment-1656376784 and on Slack channel, a password strength checker is helpful to users when it can provide recommendations to make their passwords stronger. Only telling users their passwords WEAK or STRONG is rather vague and not actionable for users. 'This is a top-100 common password' is also in the same line of vagueness.

Instead of 'This is a top-100 common password', I would suggest we display recommendations such as:
"We recommend using passwords with random combinations of upper and lowercase letters, numbers, and special characters."
This way, it guides the users on what to do to make their password stronger.

Comment (DarshitChanpura) [reply public/apps/configuration/utils/test/password-strength-bar.test.tsx:34]: @derek-ho, based on @kamingleung's comments. I think this block: https://github.com/opensearch-project/security-dashboards-plugin/pull/1762/files#diff-ac97f3618a6ad790d8d2692a16dcd87421f4aee2de3f856167cd740505d31b52R60 might not be necessary. Instead, can we maintain a separate list of feedback messages so we don't rely on the ones coming from ZXCVBN for a more uniform user experience?

**Review by DarshitChanpura — COMMENTED** [review]
Looks good. From what I see, all the passwords with score 0 or 1 or 2 will receive a generic message: `Add another word or two. Uncommon words are better.` to suggest improvements. And passwords with score of 3 or 4 will not see any message. @derek-ho Could you replace the video in the PR description with this new look-and-feel?

**Comment by derek-ho** [comment]
> Looks good. From what I see, all the passwords with score 0 or 1 or 2 will receive a generic message: `Add another word or two. Uncommon words are better.` to suggest improvements. And passwords with score of 3 or 4 will not see any message. @derek-ho Could you replace the video in the PR description with this new look-and-feel?

Correct - updated the video!

**Comment by kamingleung** [comment]
@derek-ho Thanks for the updates.

When the password is weak or very weak, we should show the text and the progress bar in red "danger" color. The green color is telling users a false sense that everything is fine.
image

Also, let's adjust the padding:
image

**Comment by derek-ho** [comment]
> @derek-ho Thanks for the updates.
> 
> When the password is weak or very weak, we should show the text and the progress bar in red "danger" color. The green color is telling users a false sense that everything is fine. When it is strong or very strong, we can show the green color. image
> 
> Also, let's adjust the padding: image

Thanks @kamingleung for the suggestions! I made the updates and here is an updated video, I think it looks a lot better, thanks! Let me know what you think.




https://github.com/opensearch-project/security-dashboards-plugin/assets/26328171/0f7d2c5b-5b7f-4cf1-99c6-b322e64ade57

**Review by DarshitChanpura — COMMENTED** [review]
The changes look good to me. Thanks for adding color to the strength bar. Waiting for CI to pass before adding my approval.

**Comment by kamingleung** [comment]
@derek-ho Thanks for updating! Is there a place we can find all of the suggestion/recommendation text? We should review the messages with our writers. cc: @vagimeli

**Comment by vagimeli** [comment]
> @derek-ho Thanks for updating! Is there a place we can find all of the suggestion/recommendation text? We should review the messages with our writers. cc: @vagimeli

If I can make changes directly in the content file, that is most time efficient. Otherwise, my changes will need to be copy/paste into the primary source file.

**Comment by derek-ho** [comment]
Unfortunately we are not coming up with the text - it is coming form a library we are using, for example within this file: https://github.com/dropbox/zxcvbn/blob/master/src/feedback.coffee#L7. Does the content need to be provided by us to be approved for use/merged?

**Comment by vagimeli** [comment]
> Unfortunately we are not coming up with the text - it is coming form a library we are using, for example within this file: https://github.com/dropbox/zxcvbn/blob/master/src/feedback.coffee#L7. Does the content need to be provided by us to be approved for use/merged?

I'm unfamiliar with this workflow. Does this mean that the UI copy/text is auto-generated? How does the text get updated? For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner. How does the library know the correct style, tone, or terminology to use?

**Comment by derek-ho** [comment]
> > Unfortunately we are not coming up with the text - it is coming form a library we are using, for example within this file: https://github.com/dropbox/zxcvbn/blob/master/src/feedback.coffee#L7. Does the content need to be provided by us to be approved for use/merged?
> 
> I'm unfamiliar with this workflow. Does this mean that the UI copy/text is auto-generated? How does the text get updated? For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner. How does the library know the correct style, tone, or terminology to use?

It means that the text/copy that is generated under the password strength bar is coming from the library/not coming from us. For example, that text/copy may be "Use a longer keyboard pattern with more turns", or "Avoid repeated words and characters".

**Comment by kamingleung** [comment]
> For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner.

@vagimeli For additional context, this issue is specifically on the password strength indicator in the form. However, you made a good point to overall improve text on this page. It may be worth we taking a pass on reviewing the page and create new issues on this.

> text/copy that is generated under the password strength bar is coming from the library/not coming from us. For example, that text/copy may be "Use a longer keyboard pattern with more turns", or "Avoid repeated words and characters".

@derek-ho Are we able to customize the messages at all?

**Comment by vagimeli** [comment]
> > For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner.
> 
> image
> @vagimeli For additional context, this issue is specifically on the password strength indicator in the form. However, you made a good point to overall improve text on this page. It may be worth we taking a pass on reviewing the page and create new issues.
> 
> > text/copy that is generated under the password strength bar is coming from the library/not coming from us. For example, that text/copy may be "Use a longer keyboard pattern with more turns", or "Avoid repeated words and characters".
> 
> @derek-ho Are we able to customize the messages at all?

@kamingleung @derek-ho Thank you for the context. My understanding is that UX/UI copy is generated using a UI strings platform. Currently, the process for writing and editing copy is a manual, time-intensive process. I'm working a proposal for a better, more organized process between devs and writers, editor (@natebower), and UX designers. I'll share with on on Monday, 4/22.

**Comment by vagimeli** [comment]
> > > For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner.
> > 
> > 
> > image
> > @vagimeli For additional context, this issue is specifically on the password strength indicator in the form. However, you made a good point to overall improve text on this page. It may be worth we taking a pass on reviewing the page and create new issues.
> > > text/copy that is generated under the password strength bar is coming from the library/not coming from us. For example, that text/copy may be "Use a longer keyboard pattern with more turns", or "Avoid repeated words and characters".
> > 
> > 
> > @derek-ho Are we able to customize the messages at all?
> 
> @kamingleung @derek-ho Thank you for the context. My understanding is that UX/UI copy is generated using a UI strings platform. Currently, the process for writing and editing copy is a manual, time-intensive process. I'm working on a proposal for a better, more organized process between devs and writers, editor (@natebower), and UX designers. I'll share with on on Monday, 4/22.

@derek-ho @kamingleung Here's my suggested rewrites for the password strength UI copy:

- Change "Ok" to be more descriptive. For example, use "Weak", "Medium", or "Strong".
- Change UI description based on the strength output. 

Suggested text for each strength: 
- Weak: "Your password needs to be stronger. Use a mix of uppercase, lowercase, numbers, and special characters."
- Medium: "Your password is getting there, but it could be stronger. Consider adding more character types."
- Strong: "Your password is secure and strong."

**Comment by derek-ho** [comment]
> > > > For example, in the Username section, the descriptive text needs to read "username" instead of "user name." Overall, UI copy needs to be written in a concise, user-friendly manner.
> > > 
> > > 
> > > image
> > > @vagimeli For additional context, this issue is specifically on the password strength indicator in the form. However, you made a good point to overall improve text on this page. It may be worth we taking a pass on reviewing the page and create new issues.
> > > > text/copy that is generated under the password strength bar is coming from the library/not coming from us. For example, that text/copy may be "Use a longer keyboard pattern with more turns", or "Avoid repeated words and characters".
> > > @derek-ho Are we able to customize the messages at all?
> > 
> > 
> > @kamingleung @derek-ho Thank you for the context. My understanding is that UX/UI copy is generated using a UI strings platform. Currently, the process for writing and editing copy is a manual, time-intensive process. I'm working on a proposal for a better, more organized process between devs and writers, editor (@natebower), and UX designers. I'll share with on on Monday, 4/22.
> 
> @derek-ho @kamingleung Here are my suggested rewrites for the password strength UI copy:
> 
> * Change "Ok" to be more descriptive. For example, use "Weak", "Medium", or "Strong". If you leave it as "Ok", change to proper spelling "OK".
> * Change UI descriptive text for each strength output. Suggested text for each strength:
> * Weak: "Your password needs to be stronger. Use a mix of uppercase, lowercase, numbers, and special characters."
> * Medium: "Your password is getting there, but it could be stronger. Consider adding more character types."
> * Strong: "Your password is secure and strong."

Thanks for the suggestions! @kamingleung we wouldn't be able to modify those texts unfortunately - it would be all or nothing - if we need to have full control/review over text that shows up in the product then we may need to remove those suggestions all together, and replace it with the generic suggestions provided by @vagimeli in the quoted reply, does that sound acceptable to everyone?

**Comment by cwperks** [comment]
@derek-ho Thank you for all of the work on this! One last thing I wanted to add is that there is a setting called `plugins.security.restapi.password_score_based_validation_strength` which has a default set to Strong. (See [PR](https://github.com/opensearch-project/security/issues/2569) for details). We could consider disabling the save button if the Password and Re-enter password are not above the minimal acceptable pw strength.

**Comment by derek-ho** [comment]
> @derek-ho Thank you for all of the work on this! One last thing I wanted to add is that there is a setting called `plugins.security.restapi.password_score_based_validation_strength` which has a default set to Strong. (See [PR](https://github.com/opensearch-project/security/issues/2569) for details). We could consider disabling the save button if the Password and Re-enter password are not above the minimal acceptable pw strength.

As this is a configurable value, what do you think of doing a follow up where we return this value in dashboards info from the backend and then disable/enable via the set value instead of always using Strong?

**Comment by derek-ho** [comment]
@cwperks @DarshitChanpura @kamingleung can you folks do a final review of this? If we are ok with having text from a library show up in the product, this PR is ready. If not, I will remove that usage and replace it with a generic message as suggested above by @vagimeli

**Review by DarshitChanpura — APPROVED** [review]
LGTM. IMO the text can be modified in the follow-up PR, and is a non-blocker.

**Review by cwperks — APPROVED** [review]
LGTM, thanks @derek-ho!

**Comment by cwperks** [comment]
@derek-ho Can you please rebase or merge the latest from main into this branch?

**Pull request merged**
